Add statistics struct and Support manually inject statistics

EmmyMiao87 opened the pull request:
## Proposed changes

This PR mainly developed the data structure used by statistical information
and the function of manually modifying the statistical information.
We use a statistics package alone to store statistical information,
and use the 'statistics manager' as a unified entry for statistical information.
For detailed data structure and explanation, please refer to the comments on the class.

Manually modify statistics include: Manually modify table statistics and column statistics.
The syntax is explained in the issue #6370.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

EmmyMiao87 committed:
[WIP] Add statistics struct and Support manually inject statistics

This PR mainly developed the data structure used by statistical information
and the function of manually modifying the statistical information.
We use a statistics package alone to store statistical information,
and use the 'statistics manager' as a unified entry for statistical information.
For detailed data structure and explanation, please refer to the comments on the class.

Manually modify statistics include: Manually modify table statistics and column statistics.
The syntax is explained in the issue #6370.

morningman commented on fe/fe-core/src/main/cup/sql_parser.cup:
```diff
@@ -827,6 +828,20 @@ alter_stmt ::=
     {:
         RESULT = new AlterRoutineLoadStmt(jobLabel, jobProperties, datasourceProperties);
     :}
+    | KW_ALTER KW_TABLE table_name:tbl KW_SET KW_STATS LPAREN key_value_map:map RPAREN
```
Could you please add this new statement in #6277 ?
I think they should be under `Database Administration Statements`(same as MySQL)

EmmyMiao87 replied:
DONE

morningman commented on fe/fe-core/src/main/cup/sql_parser.cup:
```diff
@@ -263,7 +263,7 @@ terminal String KW_ADD, KW_ADMIN, KW_AFTER, KW_AGGREGATE, KW_ALIAS, KW_ALL, KW_A
     KW_S3, KW_SCHEMA, KW_SCHEMAS, KW_SECOND, KW_SELECT, KW_SEMI, KW_SERIALIZABLE, KW_SESSION, KW_SET, KW_SETS, KW_SET_VAR, KW_SHOW, KW_SIGNED,
     KW_SKEW,
-    KW_SMALLINT, KW_SNAPSHOT, KW_SONAME, KW_SPLIT, KW_START, KW_STATUS, KW_STOP, KW_STORAGE, KW_STREAM, KW_STRING, KW_STRUCT,
+    KW_SMALLINT, KW_SNAPSHOT, KW_SONAME, KW_SPLIT, KW_START, KW_STATUS, KW_STATS, KW_STOP, KW_STORAGE, KW_STREAM, KW_STRING, KW_STRUCT,
```
Add KW_STATS to `keywords`?

EmmyMiao87 replied:
done

morningman commented on fe/fe-core/src/main/java/org/apache/doris/statistics/StatisticsManager.java:
```diff
@@ -0,0 +1,70 @@
+            ErrorReport.reportDdlException(ErrorCode.ERR_BAD_FIELD_ERROR, columnName);
+        }
+        // match type and column value
+        statistics.updateColumnStats(table.getId(), columnName, column.getType(), stmt.getProperties());
```
No need lock?

EmmyMiao87 replied:
use concurrent map instead

morningman commented on fe/fe-core/src/main/java/org/apache/doris/statistics/OlapTableStats.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class OlapTableStats extends TableStats {
+
+    private Map<Long, TabletStats> idToTabletStats = Maps.newHashMap();
```
How about put this to TabletInvertedIndex?

EmmyMiao87 replied:
All classes in statistical information are currently designed to be completely decoupled from other meta-information.

morningman reviewed (dismissed):
LGTM

EmmyMiao87 committed:
Show table and column statistics

'SHOW TABLE STATS' used to show the statistics of table.
'SHOW COLUMN STATS' used to show the statistics of columns.

Currently, only the tables and columns for setting statistics
will be displayed in the results.

morningman reviewed (approved):
LGTM
